### PR TITLE
Solving fastq/fastq_screen problem when running same sample in different FC but same lane

### PR DESCRIPTION
**Pull request by vezzi**
I am still actively working, this is only to see what solution I have in mind and see if everyone agrees.

I solved only for fastqc the problem of not being able to run fastqc when the same sample is run on different FC but on the same lane number.

Parser.py instead has been modified to solve a different but related bug: if you have the same sample sequenced in FC1_lane1, FC1_lane2, FC2_lane1, and FC2_lane2 then fastqscreen is not run at all
this because the code responsible to parse the pairs instead of producing a list of 4 couples produces  2 lists with 4 files (2 couples). --> this problem is now solved

### CODE REVIEW

**Comment by vezzi** [comment]
@Galithil changes done. fastqc should be fine, but I have yet to test it extensively. I will now start to work on fastq_screen

**Comment by vezzi** [comment]
@Galithil @b97pla can you please review this PR. SOlves the problem of collisions in naming in QC.
Need still to be tested by @sylvinite so DO NOT MERGE

**Comment by sylvinite** [comment]
Ping me when you're done and I'll pump it into a staging release on irma.

File: ngi_pipeline/engines/qc_ngi/workflows.py [hunk 1]
```diff
@@ -85,14 +85,15 @@ def workflow_fastqc(input_files, output_dir, config):
     fastqc_output_file_tmpls = ("{}_fastqc.zip", "{}_fastqc.html")
     fastq_to_analyze = set()
     for fastq_file in fastq_files:
-        # Get the basename withot extensions (.fastq, .fastq.gz)
         m = re.match(r'([\w-]+).fastq', os.path.basename(fastq_file))
+        #fetch the FCid
+        fc_id = re.search('\d{4,6}_(?P<fcid>[A-Z0-9]{10})', fastq_file).groups()[0]
```
Comment (b97pla) [inline ngi_pipeline/engines/qc_ngi/workflows.py]: isn't this regexp both a bit too specific and unspecific? how about catching on either the entire runfolder name, e.g. `\d{6}_[A-Z0-9-]+_\d{4,}_(?P<fcid>[A-Z0-9]{10})` or relying on the organization and get the fcid from the parent folder, e.g. `os.path.dirname(fastq_file).split("_")[-1]`?

File: ngi_pipeline/engines/qc_ngi/workflows.py [hunk 2]
```diff
@@ -103,21 +104,39 @@ def workflow_fastqc(input_files, output_dir, config):
...
-        # Module loading
-        modules_to_load = get_all_modules_for_workflow("fastqc", config)
-        for module in modules_to_load:
-            cl_list.append("module load {}".format(module))
-        # Execute fastqc
+    # fastqc commands
+    for fastq_file in fastq_to_analyze:
+        #when building the fastqc command soflink in the qc_ngi folder the fastq file processed being sure to avoid name collision (i.e., same sample run in two different FC but on the same lane number). Run fastqc on the softlink and delete the soflink straight away. In case the fastq file does not match the regex force the exectuion of fastqc... this is what we have been always doing.
+        fc_id = re.search('\d{4,6}_(?P<fcid>[A-Z0-9]{10})', fastq_file).groups()[0]
+        m = re.match(r'([\w-]+).(fastq.*)', os.path.basename(fastq_file))
```
Comment (b97pla) [inline ngi_pipeline/engines/qc_ngi/workflows.py]: This is a repetition of the same parsing/logic as above. I would prefer to put this in a function or variable in order not to repeat the logic.

File: ngi_pipeline/engines/qc_ngi/workflows.py [hunk 3]
```diff
@@ -153,53 +172,58 @@ def workflow_fastq_screen(input_files, output_dir, config):
...
     fastq_to_analyze = set()
-    for elt in input_files:
-        # This may be a read pair
-        if type(elt) is list:
-            # Changing list to tuple so we can use it in the set() (lists aren't hashable)
-            elt = tuple(elt)
-            # fastq_screen uses the name of the first read of the pair for output files
-            fastq_file = elt[0]
+    for fastq_file in fastq_files:
+        m = re.match(r'([\w-]+).fastq', os.path.basename(fastq_file))
```
Comment (b97pla) [inline ngi_pipeline/engines/qc_ngi/workflows.py]: this code looks very similar to the code above.. I think a refactoring to remove the redundancy would be helpful

**Comment by vezzi** [comment]
thanks @b97pla for the comments. I implemented them and I am testing it now. There is still some code repetition when I create the fastqc and fastq_screen commands, but given that the commands are different I think it is acceptable.

**Comment by b97pla** [comment]
:+1: looks good

**Comment by sylvinite** [comment]
Verified to successfully work with ngi_pipeline, taca deliveries and multiqc.

**Comment by vezzi** [comment]
@b97pla we are going to merge this and deploy next time on irma if it is fine with you. We have done testing and we are happy with it.

**Comment by b97pla** [comment]
ok, nice work :+1:

**Pull request merged**